contracts/whitelists/whitelist-merkletree/src/contract.rs

```diff
@@ -0,0 +1,288 @@
+
+    Ok(Response::new()
+        .add_attribute("action", "instantiate")
+        .add_attribute("contract_name", CONTRACT_NAME)
+        .add_attribute("contract_version", CONTRACT_VERSION)
+        .add_attribute("sender", info.sender))
+}
+
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::UpdateStartTime(time) => execute_update_start_time(deps, env, info, time),
+        ExecuteMsg::UpdateEndTime(time) => execute_update_end_time(deps, env, info, time),
+        ExecuteMsg::UpdateAdmins { admins } => execute_update_admins(deps, env, info, admins),
+        ExecuteMsg::Freeze {} => execute_freeze(deps, env, info),
+    }
+}
+
+pub fn execute_update_start_time(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    start_time: Timestamp,
+) -> Result<Response, ContractError> {
+    let mut config = CONFIG.load(deps.storage)?;
+    can_execute(&deps, info.sender.clone())?;
+
+    // don't allow updating start time if whitelist is active
+    if env.block.time >= config.start_time {
+        return Err(ContractError::AlreadyStarted {});
+    }
+
+    if start_time > config.end_time {
+        return Err(ContractError::InvalidStartTime(start_time, config.end_time));
+    }
+
+    let genesis_start_time = Timestamp::from_nanos(GENESIS_MINT_START_TIME);
+    let start_time = if start_time < genesis_start_time {
+        genesis_start_time
+    } else {
+        start_time
+    };
+
+    config.start_time = start_time;
+    CONFIG.save(deps.storage, &config)?;
+    Ok(Response::new()
+        .add_attribute("action", "update_start_time")
+        .add_attribute("start_time", start_time.to_string())
+        .add_attribute("sender", info.sender))
+}
+
+
+pub fn execute_update_end_time(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    end_time: Timestamp,
+) -> Result<Response, ContractError> {
+    let mut config = CONFIG.load(deps.storage)?;
+    can_execute(&deps, info.sender.clone())?;
+
+    if env.block.time >= config.start_time && end_time > config.end_time {
+        return Err(ContractError::AlreadyStarted {});
+    }
+
+    if end_time < config.start_time {
+        return Err(ContractError::InvalidEndTime(end_time, config.start_time));
+    }
+
+    config.end_time = end_time;
+    CONFIG.save(deps.storage, &config)?;
+    Ok(Response::new()
+        .add_attribute("action", "update_end_time")
+        .add_attribute("end_time", end_time.to_string())
+        .add_attribute("sender", info.sender))
+}
+
+fn store_tree_uri(
+    deps: DepsMut,
+    tree_uri: Option<String>
+) -> Result<(), ContractError> {
+    if tree_uri.is_some() {
+        let parsed_token_uri = Url::parse(&tree_uri.as_ref().unwrap())
+                    .map_err(|_| ContractError::InvalidMerkleTreeURI {})?;
+
+        if parsed_token_uri.scheme() != "ipfs" {
```

Comment:
I don't think we need an IPFS requirement here as whitelists are usually ephemeral.

Reply:
Valid point. Added a commit removing the schema enforcing part